### PR TITLE
Fixes #16477: Update relayd to tokio 0.2

### PR DESCRIPTION
**Pull request by amousset**
https://issues.rudder.io/issues/16477

### CODE REVIEW

**Comment by amousset** [comment]
Commit modified

**Comment by amousset** [comment]
Still a WIP (some cleaning to do) but it's close.

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: relay/sources/relayd/src/data/shared_file.rs [hunk 1]
```diff
@@ -141,7 +143,8 @@ impl FromStr for Metadata {
         for line in s.lines() {
             let kv: Vec<&str> = line.splitn(2, '=').collect();
             if kv.len() == 2 && parsed.insert(kv[0], kv[1]).is_some() {
-                return Err(Error::DuplicateHeader(line.to_string()));
+                let e: Error = RudderError::DuplicateHeader(line.to_string()).into();
```
Comment (gpoblon) [inline relay/sources/relayd/src/data/shared_file.rs:146]: Is this error variable of any use?

File: relay/sources/relayd/src/data/shared_file.rs [hunk 2]
```diff
@@ -163,7 +166,7 @@ impl FromStr for Metadata {
 
         let digest = extract(&parsed, "digest")?.to_string();
         // Validate hexadecimal string
-        hex::decode(&digest).map_err(|_| Error::InvalidHeader(digest.clone()))?;
+        hex::decode(&digest).map_err(|_| RudderError::InvalidHeader(digest.clone()))?;
```
Comment (gpoblon) [inline relay/sources/relayd/src/data/shared_file.rs:169]: why isnt "into()" method required here?

File: relay/sources/relayd/src/output/upstream.rs [hunk 3]
```diff
@@ -38,36 +36,39 @@ pub fn send_inventory(
...
+        ))
+        .basic_auth(
+            &job_config.cfg.output.upstream.user,
+            Some(&password.value()),
+        )
+        .body(Body::from(content))
+        .send()
+        .await;
+
+    result
```
Comment (gpoblon) [inline relay/sources/relayd/src/output/upstream.rs:68]: Maybe you'll want to have Err handled here

**Comment by amousset** [comment]
PR updated with a new commit

**Comment by amousset** [comment]
PR updated with a new commit

File: relay/sources/relayd/src/api/shared_folder.rs [hunk 4]
```diff
@@ -1,13 +1,56 @@
...
+
+    let job_config_head = job_config.clone();
+    let head = method::head()
+        .and(base)
+        .map(move || job_config_head.clone())
+        .and(path::peek())
+        .and(query::<SharedFolderParams>())
+        .and_then(|j, p, q| handlers::head(p, q, j));
+
+    let job_config_get = job_config;
```
Comment (gpoblon) [inline relay/sources/relayd/src/api/shared_folder.rs:25]: does this variable have a purpose?

File: relay/sources/relayd/src/output/upstream.rs (hunk 3 above)
Comment (amousset) [reply relay/sources/relayd/src/output/upstream.rs:68]: well spotted, fixed!

File: relay/sources/relayd/src/api/shared_folder.rs (hunk 4 above)
Comment (amousset) [reply relay/sources/relayd/src/api/shared_folder.rs:25]: not this one indeed, as it is the last one and it is not moved by a closure.

**Review by gpoblon — APPROVED** [review]

**Comment by amousset** [comment]
OK, squash merging this PR

**Pull request merged**